packages/tools/devtools/devtools-view/src/components/TelemetryView.tsx

```diff
@@ -96,31 +100,76 @@ export function TelemetryView(): React.ReactElement {
 		};
 	}, [messageRelay, setTelemetryEvents]);
 
+	React.useEffect(() => {
+		if (
+			telemetryEvents &&
+			bufferedEvents.length > 0 &&
+			telemetryEvents.length < maxEventsToDisplay
+		) {
+			const newEvents = bufferedEvents.slice(0, maxEventsToDisplay - telemetryEvents.length);
+			const remainingBuffer = bufferedEvents.slice(
+				maxEventsToDisplay - telemetryEvents.length,
+			);
+			setTelemetryEvents([...newEvents, ...telemetryEvents]);
+			setSelectedIndex((prevIndex) => prevIndex + newEvents.length);
+			setBufferedEvents(remainingBuffer);
+		}
+	}, [telemetryEvents, bufferedEvents, maxEventsToDisplay, selectedIndex]);
```

Comment:
Note: because you observe the value of `telemetryEvents` and `bufferedEvents` here, the Effect will re-run any time either of the values changes. That means we will be re-setting all of the `telemetryEvents` values any time `bufferedEvents` changes, for example, which will also force us to re-render all of the contents.

I would recommend trying to split up the state management here. This could be done by using separate Effects for the buffer and `telemetryEvents`. It could also be done by splitting this into multiple components.

Reply:
I thought about this a bit more and I don't think this separation is necessary or even a good thing. While it's true that the useEffect will run whenever `telemetryEvents` or `bufferedEvents` changes, re-renders aren't caused by the hook itself, but by the setters. Furthermore `telemetryEvents` & `bufferedEvents` are closely related and should be coordinated to prevent issues like duplication or event loss. 

Most importantly, the conditions in the useEffect determine whether these setters will run, causing a re-render. So, if there are no events in the telemetryEvents or bufferedEvents array, or if the number of telemetryEvents is already equal to or exceeds maxEventsToDisplay, the code block inside the if statement won't run.

This happens often when the component first loads (no telemetry data yet), or when no more data can be displayed because it's already at the maxEventsToDisplay limit. Hope that made sense @Josmithr 

CC: @alexvy86

Reply:
Yeah, that made sense. And it's a good callout that it isn't the hook re-running that causes re-renders, it's the underlying state changes. Those are easy to conflate.